src/glmfit.jl

```diff
@@ -290,48 +295,54 @@ function nulldeviance(m::GeneralizedLinearModel)
     return dev
 end
 
-function loglikelihood(m::AbstractGLM)
-    r   = m.rr
-    wts = r.wts
-    y   = r.y
-    mu  = r.mu
-    d   = r.d
-    ll  = zero(eltype(mu))
-    if !isempty(wts)
-        ϕ = deviance(m)/sum(wts)
-        @inbounds for i in eachindex(y, mu, wts)
+loglikelihood(m::AbstractGLM) = loglikelihood(m.rr)
+
+
+function loglikelihood(r::GlmResp{T,D,L,<:AbstractWeights}) where {T,D,L}
+    y = r.y
+    mu = r.mu
+    wts = weights(r)
+    sumwt = sum(wts)
+    d = r.d
+    ll = zero(eltype(mu))
+    n = nobs(r)
+    N = length(y)
+    δ = deviance(r)
+    ϕ = δ/n
+    if wts isa FrequencyWeights || wts isa UnitWeights
+        @inbounds for i in eachindex(y, mu)
             ll += loglik_obs(d, y[i], mu[i], wts[i], ϕ)
         end
     else
-        ϕ = deviance(m)/length(y)
-        @inbounds for i in eachindex(y, mu)
-            ll += loglik_obs(d, y[i], mu[i], 1, ϕ)
-        end
+        @inbounds for i in eachindex(y, mu, wts)
+            ll += loglik_apweights_obs(d, y[i], mu[i], wts[i], δ, sumwt, N)
+        end       
     end
-    ll
+    return ll
 end
 
 function nullloglikelihood(m::GeneralizedLinearModel)
     r      = m.rr
-    wts    = r.wts
+    wts    = weights(m)
+    sumwt  = sum(wts)
     y      = r.y
     d      = r.d
     offset = r.offset
     hasint = hasintercept(m)
-    ll  = zero(eltype(y))
+    ll     = zero(eltype(y))
     if isempty(r.offset) # Faster method
-        if !isempty(wts)
-            mu = hasint ? mean(y, weights(wts)) : linkinv(r.link, zero(ll)/1)
-            ϕ = nulldeviance(m)/sum(wts)
+        mu = hasint ? mean(y, wts) : linkinv(r.link, zero(ll)/1)
+        δ = nulldeviance(m)
+        ϕ = nulldeviance(m)/nobs(m)
+        N = length(y)
+        if wts isa FrequencyWeights || wts isa UnitWeights
```

Comment:
Better either be consistent and either use `wts isa Union{...}` everywhere or `wts isa ... || wts isa ...` everywhere.